DQM/L1TMonitor/src/L1TStage2CaloLayer1.cc

```diff
@@ -29,14 +30,50 @@ L1TStage2CaloLayer1::L1TStage2CaloLayer1(const edm::ParameterSet& ps)
       fedRawData_(consumes<FEDRawDataCollection>(ps.getParameter<edm::InputTag>("fedRawDataLabel"))),
       histFolder_(ps.getParameter<std::string>("histFolder")),
       tpFillThreshold_(ps.getUntrackedParameter<int>("etDistributionsFillThreshold", 0)),
-      ignoreHFfbs_(ps.getUntrackedParameter<bool>("ignoreHFfbs", false)) {
-  ecalTPSentRecd_.reserve(28 * 2 * 72);
-  hcalTPSentRecd_.reserve(41 * 2 * 72);
-}
+      ignoreHFfbs_(ps.getUntrackedParameter<bool>("ignoreHFfbs", false)) {}
 
 L1TStage2CaloLayer1::~L1TStage2CaloLayer1() {}
 
-void L1TStage2CaloLayer1::analyze(const edm::Event& event, const edm::EventSetup& es) {
+//utility function for updating the link event and max error mapping
+//input: event, one of the error maps, and the corresponding type of errors per event
+//output: none, but changes the error map if the number of errors is greater than the one stored
+//for the current luminosity block
+void L1TStage2CaloLayer1::updateMaxErrorMapping(const edm::Event& event,
+                                                const std::unique_ptr<std::map<std::string, int>>& theMap,
+                                                const int nErrors) const {
+  std::string lumiBlock = std::to_string(event.id().luminosityBlock());
+  if (theMap->find(lumiBlock) != theMap->end()) {
+    //this is not the first event with errors per lumi-block, insert it only if the number of errors is larger than before
+    if (nErrors > theMap->find(lumiBlock)->second) {
+      (*theMap)[lumiBlock] = nErrors;
```

Comment:
This approach is not thread safe (I'm saying it only here and not repeat everywhere). The compiler would also complain if the map would be taken as `const std::map<std::string, int>&` (unfortunately "const pointer" is not the same as "(const) pointer to const", which is what would be really needed here).

If I understood correctly, the aim is to find a maximum of some quantity over a LuminosityBlock, and eventually put that number into a bin of a histogram that corresponds that lumi. I believe the most efficiency way to do that (in general) would be to (following the example https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#example_usage_calculating_per_lu)
* Use [`edm::StreamCache`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_StreamCacheT) with an `int` for each quantity to be recorded over a lumi
  * In `dqmAnalyze()` just do along `streamCache(event.streamID).var = std::max(streamCache(event.streamID).var, localVar`) (*)
* Use [`edm::LuminosityBlockSummaryCache`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_LuminosityBlockSummaryCacheT) with an `int` for each quantity to do a reduction (with `std::max()` operation) over the edm streams
   * `streamEndLuminosityBlockSummary()` stores the maximum of the StreamCache and the LumiSummaryCache to the LumiSummaryCache and resets the StreamCache value(s) to zero
  * `globalEndLuminosityBlockSummary()` stores the reduced maximum into a container (see next bullet) along with the lumi id
* Use `tbb::concurrent_vector<std::pair<edm::LuminosityBlockNumber_t, int>>` in the `RunCache` (i.e. in `CaloL1Information::monitoringDataHolder`) to hold the values per lumi for that run
   * `dqmEndRun()` copies the values from the container into the histogram

(alternatively a [`edm::LuminosityBlockCache`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_LuminosityBlockCacheT) could be combined with `std::atomic<int>`s)

(*) for @cms-sw/dqm-l2, the `DQMGlobalEDAnalyzer` API would be more consistent with `edm::global` interfaces if the `edm::StreamID` would be passed from `accumulate()` to `dqmAnalyze()`

Reply:
@makortel Thanks for the comments. I was afraid this might be the case.

> If I understood correctly, the aim is to find a maximum of some quantity over a LuminosityBlock, and eventually put that number into a bin of a histogram that corresponds that lumi

This is correct. Unfortunately, I think I've tried to use `LuminosityBlockCaches` to gain access to the luminosity block transition functions and cache information in this manner (the old implementation of this module was designed to fill things on luminosity block transitions, this seemed like the quickest way to regain all functionality). My attempt was something similar to:

`class L1TStage2CaloLayer1Offline : public DQMGlobalEDAnalyzer<edm::LuminosityBlockCache<CaloL1Information::monitoringDataHolder>>`

as the main implementation of the module, but his doesn't work due to the way DQM global are implemented, as run level caching is assumed at the dqm global analyzer level here:

https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/interface/DQMGlobalEDAnalyzer.h#L13-L19

So I think this rules out every solution you mentioned but the last. However, I am unfamiliar with `tbb::concurrent_vector<>`'s. Is there documentation I should be aware of for using these? I assume these objects are just a vector with some sort of inherent locking implemented?

I had also considered an even more "terse" solution, comparing and setting in the monitoring element directly for every event, I assume this would be no more thread safe than what I have implemented right now?

Thanks again for your help,
Andrew Loeliger

Reply:
> My attempt was something similar to:
> 
> `class L1TStage2CaloLayer1Offline : public DQMGlobalEDAnalyzer<edm::LuminosityBlockCache<CaloL1Information::monitoringDataHolder>>`
> 
> as the main implementation of the module, but his doesn't work due to the way DQM global are implemented, as run level caching is assumed at the dqm global analyzer level here:
> 
> https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/interface/DQMGlobalEDAnalyzer.h#L13-L19

The `DQMGlobalEDAnalyzer` indeed requires/uses a RunCache for the MonitorElements, but (as far as I can tell) that does not preclude the use of other caches for other information (especially when the functions specific to those caches do not use any DQM APIs).

> So I think this rules out every solution you mentioned but the last. 

I meant that all the three bullet points need to be used together.

The alternative I suggested in the parentheses under the bullet points would replace the use of `edm::StreamCache` and `edm::LuminosityBlockSummaryCache` with `edm::LuminosityBlockCache` and explicit use of `std::atomic<int>` (which would require a somewhat-low-level loop using `std::atomic<int>::compare_exchange_weak()` to achieve the `max()` operation).

> However, I am unfamiliar with `tbb::concurrent_vector<>`'s. Is there documentation I should be aware of for using these? I assume these objects are just a vector with some sort of inherent locking implemented?

It provides fine-grained synchronization (and thus might be a bit overkill for just concurrent `push_back()`). The interface is very `std::vector`-like (including `push_back()` and iteration over elements). For more information please see
https://www.threadingbuildingblocks.org/docs/help/tbb_userguide/concurrent_vector_ug.html
https://www.threadingbuildingblocks.org/docs/doxygen/a00046.html

> I had also considered an even less verbose solution, comparing and setting in the monitoring element directly for every event, I assume this would be no more thread safe than what I have implemented right now?

Right, not without a per-histogram locking (which the MonitorElement actually has, but is not exposed outside, for good reasons).

Reply:
>The DQMGlobalEDAnalyzer indeed requires/uses a RunCache for the MonitorElements, but (as far as I can tell) that does not preclude the use of other caches for other information (especially when the functions specific to those caches do not use any DQM APIs).

Then, can I ask what is the appropriate syntax to use when trying to mix and use the different cache scopes at various levels? As I mentioned, I tried the above syntax:

```
class L1TStage2CaloLayer1Offline : public DQMGlobalEDAnalyzer<edm::LuminosityBlockCache<CaloL1Information::monitoringDataHolder>>
```
but later, when trying to retrieve this luminosity block cache doing something like:

```
if (nEcaLLinkErrors > luminosityBlockCache(event.getLuminosityBlock().index())->maxEvtLinkErrorsECALCurrentLumi_)
```

This would always error in the compiler as it didn't recognize the `luminosityBlockCache()` function for retrieving the luminosity block cache (it is also perhaps worth noting that it would also not recognize overrides of the global luminosity block transition functions, stating the override was not overriding anything previously defined).

This error was unique to the `DQMGlobalEDAnalyzer` implementation, when changing from `DQMGlobalEDAnalyzer` to `DQMOneEDAnalyzer` these lines passed the compiler without complaint. 

Is there a method I am unaware of to use these two different cache scopes together?

Thanks,
Andrew Loeliger

Reply:
This one
> ```cpp
> class L1TStage2CaloLayer1Offline : public DQMGlobalEDAnalyzer<edm::LuminosityBlockCache<CaloL1Information::monitoringDataHolder>>
> ```
declares to use `edm::LuminosityBlockCache<CaloL1Information::monitoringDataHolder>` as the RunCache (which I fully believe can lead to compilation errors`). In order to use multiple caches, each of them needs to be declared separately, e.g.
```cpp
// note CMS naming convention is start namespaces in lower case, and classes/structs in upper case
namespace caloL1Information {
  struct MonitoringDataHolder {
    // MonitorElements;
    tbb::concurrent_vector<std::pair<edm::LuminosityBlockID, int>> maxEvtLinkErrorsMapECAL;
    ...
  };

  struct PerStreamMax {
    int maxEvtLinkErrorsMapECAL = 0;
    ...
  };

  struct PerLumiMax {
    int maxEvtLinkErrorsMapECAL = 0;
    ...
  };
}

class L1TStage2CaloLayer1Offline : public DQMGlobalEDAnalyzer<
  caloL1Information::MonitoringDataHolder, // RunCache, not decorated with edm::RunCache<...> because the base class does that
  edm::StreamCache<caloL1Information::PerStreamMax>,
  edm::LuminosityBlockSummaryCache<caloL1Information::PerLumiMax>>

std::shared_ptr<caloL1Information::PerStreamMax> beginStream(edm::StreamID) const {
  return std::make_shared<caloL1Information::PerStreamMax>();
}

std::shared_ptr<caloL1Information::PerLumiMax> globalBeginLuminosityBlockSummary(edm::LuminosityBlock const&, edm::EventSetup const& ) const {
  return std::make_shared<caloL1Information::PerLumiMax>();
}

// reduction over events within one stream (and by definition within one lumi)
// there may be concurrent calls to this function, but framework guarantees that there is at most one call for any single stream
// (and therefore StreamCache can be used safely without explicit synchronization)
void dqmAnalyze(edm::Event const& iEvent, edm::EventSetup const&, caloL1Information::MonitoringDataHolder const&) const {
  ...
  auto cache = streamCache(iEvent.streamID());
  cache->maxEvtLinkErrorsMapECAL = std::max(cache->maxEvtLinkErrorsMapECAL, nEcalLinkErrors);
  ...
}

// reduction over streams
// framework guarantees that this function is called one at a time
// (and therefore StreamCache and LuminosityBlockSummaryCache can be used without explicit synchronization)
void streamEndLuminosityBlockSummary(edm::StreamID streamID, edm::LuminosityBlock const&,                                      edm::EventSetup const&, caloL1Information::PerLumiMax* lumiSummaryCache) const {
  auto cache = streamCache(streamID);
  lumiSummaryCache->maxEvtLinkErrorsMapECAL = std::max(lumiSummaryCache->maxEvtLinkErrorsMapECAL, cache->maxEvtLinkErrorsMapECAL);
  ...
}

// store reduced maximum to tbb::concurrent_vector
void globalEndLuminosityBlockSummary(edm::LuminosityBlock const& iLumi, edm::EventSetup const&, caloL1Information::PerLumiMax* lumiSummaryCache) const {
  auto cache = runCache(iLumi.getRun().index());
  cache->maxEvtLinkErrorsMapECAL.emplace_back(iLumi.id(), lumiSummaryCache->maxEvtLinkErrorsMapECAL);
  ...
}
```
(please invent better names for the cache types :) )